decisions/0004-remixing-react-router.md

```diff
@@ -0,0 +1,201 @@
+  return defaultShouldRevalidate;
+}
+```
+
+### `<ScrollRestoration getKey>` prop
+
+In Remix, the `<ScrollRestoration>` component made an assumption that we would always restore scroll position based on `location.key`. If the key was the same as a prior location we knew the scroll position for, then we knew you had been there before and we should restore. This works great for back/forward navigations but it's a bit overly restrictive. You cannot choose to restore scroll based on anything other than `key`. Twitter has a great implementation of this as you click around in their left nav bar - your tweet feed is always at the same place when you click back to it - even though it's a _new_ location in the history stack. This is because they're restoring by pathname here instead of `location.key`. Or maybe you want to maintain scroll position for all routes under a given pathname and you thus want to use a portion of the pathname as the scroll restoration key.
+
+In React Router we now accept an optional `<ScrollRestoration getKey>` prop where you provide a function that returns the key to use for scroll restoration:
+
+```js
+function getKey(location: Location, matches: DataRouteMatch[]) {
+  // Restore by pathname on /tweets
+  if (location.pathname === "/tweets") {
+    return location.pathname;
+  }
+  // Otherwise use the key
+  return location.key;
+}
+```
+
+**Backwards Compatibility**
+
+We're ok here since the new prop is optional and defaults to using `location.key`
+
+### `<Link resetScroll="false">` prop
+
+In addition to `<ScrollRestoration>` handling "restoring" scroll position on previously visited routes. It also handles "resetting" scroll position back to the top on _new_ routes. This is not always desirable if you're clicking around inside a tabbed view or something, so we've introduced a new `<Link resetScroll>` prop that lets you disable the scroll reset behavior _for a given navigation_. Note that this "resetting" logic happens if and only if we cannot restore scroll to a previously known location for that scroll restoration key.
+
+### `useRevalidator()` hook
+
+This has been a long time coming - see https://github.com/remix-run/remix/discussions/1996 🙂
+
+### No distinction between Error and Catch boundaries
+
+**TODO:** Ryan to fill in some more details here.
+
+The differentiation between error and catch prove to be a bit vague over time and a source of confusion for developers. We chose to go with just a single `errorElement` in the router for simplicity. If you throw anything, it ends up in the error boundary and propagates accordingly.
+
+**Backwards Compatibility**
+
+We have a few options here. In all cases, Remix v1 will provide an `errorElement` implementation that will need to do some forking to maintain backwards compatibility.
+
+1. We could introduce a new `ErrorElement` in Remix v1 and deprecate `ErrorBoundary`/`CatchBoundary` (and eventually drop them in v2)
```

Comment:
Proposal: ErrorComponent instead of ErrorElement, because you should export a component in Remix, not an element, like with ErrorBoundary/CatchBoundary/default which are all components.

Reply:
I wonder if they use `element` here to match React Router `<Route path="" element={}/>`

Reply:
yeah this was a very quick "let's call it what we call it in react router," but I agree with Sergio that from a typing standpoint you're exporting a Component in Remix, not something that's been through `createElement` 👍